File: frontend/services/vision.ts
import { supabase } from './supabase';

// Type declaration for process.env
declare const process: {
  env: {
    NEXT_PUBLIC_API_URL?: string;
    [key: string]: string | undefined;
  };
};

// API Base URL
const API_URL = typeof window !== 'undefined'
  ? (process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000')
  : 'http://backend:8000';

// Output target platforms
export type OutputTarget = 'vercel_v0' | 'claude_code' | 'cursor' | 'generic';
export type Framework = 'react' | 'vue' | 'svelte' | 'html';
export type Styling = 'tailwind' | 'css' | 'styled-components' | 'scss';
export type DetailLevel = 'basic' | 'detailed' | 'comprehensive';
export type ImageType = 'png' | 'jpg' | 'jpeg' | 'webp' | 'gif';
export type Provider = 'anthropic' | 'openai';

// Design Analysis types
export interface TypographySpec {
  classes: string;
  color: string;
}

export interface DesignAnalysis {
  layout_type: string;
  container_specs: string;
  grid_config: string | null;
  spacing_system: string;
  typography: Record<string, TypographySpec>;
  colors: Record<string, string>;
  components: string[];
  design_patterns: string[];
  interactions: string[];
}

// Prompt Package types
export interface PromptPackage {
  recommended_platform: OutputTarget;
  recommended_model: string;
  temperature: number;
  max_tokens: number;
  foundation_prompt: string;
  refinement_prompt: string | null;
  interaction_prompt: string | null;
  design_pattern_name: string;
  similar_examples: string[];
  common_pitfalls: string[];
  framework_notes: string;
}

// Request/Response types
export interface ImageToPromptRequest {
  image_base64: string;
  image_type?: ImageType;
  target_platform?: OutputTarget;
  framework?: Framework;
  styling?: Styling;
  include_interactions?: boolean;
  detail_level?: DetailLevel;
}

export interface ImageToPromptResponse {
  success: boolean;
  analysis: DesignAnalysis;
  prompt_package: PromptPackage;
  raw_markdown: string;
  usage: {
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
  };
}

export interface AnalyzeDesignRequest {
  image_base64: string;
  image_type?: ImageType;
}

export interface AnalyzeDesignResponse {
  success: boolean;
  analysis: DesignAnalysis;
  usage: {
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
  };
}

export interface PlatformInfo {
  id: OutputTarget;
  name: string;
  description: string;
  recommended_model: string;
  features: string[];
}

export interface DetailLevelInfo {
  id: DetailLevel;
  tokens: string;
  description: string;
}

export interface PlatformsResponse {
  platforms: PlatformInfo[];
  frameworks: Framework[];
  styling: Styling[];
  detail_levels: DetailLevelInfo[];
}

// Helper to get authentication token
async function getAuthToken() {
  const { data: { session } } = await supabase.auth.getSession();
  return session?.access_token;
}

// Helper to make authenticated API requests
async function apiRequest<T>(endpoint: string, method: string, body?: unknown): Promise<T> {
  const token = await getAuthToken();

  if (!token) {
    throw new Error('Authentication required');
  }

  try {
    console.log(`Making vision request to ${API_URL}${endpoint}`);

    const response = await fetch(`${API_URL}${endpoint}`, {
      method,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/json',
      },
      mode: 'cors',
      body: body ? JSON.stringify(body) : undefined,
    });

    console.log(`Response status: ${response.status}`);

    if (!response.ok) {
      let errorMessage = `API request failed with status ${response.status}`;
      try {
        const errorData = await response.json();
        errorMessage = errorData.detail || errorMessage;
      } catch {
        // If we can't parse the error as JSON, just use the status message
      }
      throw new Error(errorMessage);
    }

    return response.json();
  } catch (error) {
    console.error('Vision API request error:', error);
    throw error;
  }
}

// Convert File to base64
export async function fileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      const result = reader.result as string;
      // Remove the data URI prefix to get just the base64 data
      const base64 = result.split(',')[1];
      resolve(base64);
    };
    reader.onerror = error => reject(error);
  });
}

// Get image type from file
export function getImageType(file: File): ImageType {
  const mimeType = file.type.toLowerCase();
  const typeMap: Record<string, ImageType> = {
    'image/png': 'png',
    'image/jpeg': 'jpg',
    'image/jpg': 'jpg',
    'image/webp': 'webp',
    'image/gif': 'gif',
  };
  return typeMap[mimeType] || 'png';
}

// Vision API functions

/**
 * Translate a design image into a comprehensive prompt package
 */
export async function translateImageToPrompt(
  request: ImageToPromptRequest,
  provider: Provider = 'anthropic'
): Promise<ImageToPromptResponse> {
  return apiRequest<ImageToPromptResponse>(
    `/api/vision/translate?provider=${provider}`,
    'POST',
    {
      image_base64: request.image_base64,
      image_type: request.image_type || 'png',
      target_platform: request.target_platform || 'generic',
      framework: request.framework || 'react',
      styling: request.styling || 'tailwind',
      include_interactions: request.include_interactions ?? true,
      detail_level: request.detail_level || 'detailed',
    }
  );
}

/**
 * Analyze a design image without generating full prompts
 */
export async function analyzeDesign(
  request: AnalyzeDesignRequest,
  provider: Provider = 'anthropic'
): Promise<AnalyzeDesignResponse> {
  return apiRequest<AnalyzeDesignResponse>(
    `/api/vision/analyze?provider=${provider}`,
    'POST',
    {
      image_base64: request.image_base64,
      image_type: request.image_type || 'png',
    }
  );
}

/**
 * Get supported platforms and configuration options
 */
export async function getSupportedPlatforms(): Promise<PlatformsResponse> {
  // This endpoint doesn't require auth
  const response = await fetch(`${API_URL}/api/vision/platforms`, {
    method: 'GET',
    headers: {
      'Accept': 'application/json',
    },
  });

  if (!response.ok) {
    throw new Error('Failed to fetch supported platforms');
  }

  return response.json();
}

/**
 * Helper to translate a file directly
 */
export async function translateFileToPrompt(
  file: File,
  options: Omit<ImageToPromptRequest, 'image_base64' | 'image_type'> = {},
  provider: Provider = 'anthropic'
): Promise<ImageToPromptResponse> {
  const base64 = await fileToBase64(file);
  const imageType = getImageType(file);

  return translateImageToPrompt(
    {
      image_base64: base64,
      image_type: imageType,
      ...options,
    },
    provider
  );
}

File: backend/app/api/endpoints/vision.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from typing import Literal
import logging

from app.services.vision.vision_service import VisionService, get_vision_service
from app.models.vision import (
    ImageToPromptRequest,
    ImageToPromptResponse,
    AnalyzeDesignRequest,
    AnalyzeDesignResponse,
    DesignAnalysis,
    OutputTarget,
)
from app.services.supabase.auth import SupabaseAuthService, get_auth_service
from app.core.config import settings

router = APIRouter()
security = HTTPBearer()
logger = logging.getLogger(__name__)


@router.post("/translate", response_model=ImageToPromptResponse)
async def translate_image_to_prompt(
    request: ImageToPromptRequest,
    provider: Literal["anthropic", "openai"] = "anthropic",
    credentials: HTTPAuthorizationCredentials = Depends(security),
    auth_service: SupabaseAuthService = Depends(get_auth_service),
):
    """
    Translate a design image into a comprehensive prompt package.

    This endpoint analyzes the uploaded design image and generates:
    - Detailed visual analysis (layout, typography, colors, components)
    - Multi-step prompt package for recreating the design
    - Copyable markdown format for easy use

    The generated prompts are optimized for the specified target platform
    (Vercel v0, Claude Code, Cursor, or generic AI assistants).
    """
    try:
        # Validate user authentication
        try:
            user = await auth_service.get_user(credentials.credentials)
            logger.info(f"User authenticated for vision translate: {user.email if hasattr(user, 'email') else 'Unknown'}")
        except Exception as auth_error:
            logger.error(f"Authentication error: {str(auth_error)}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Authentication failed: {str(auth_error)}",
                headers={"WWW-Authenticate": "Bearer"},
            )

        # Validate image data
        if not request.image_base64:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Image data is required",
            )

        # Check base64 size (rough estimate: 10MB limit)
        base64_size = len(request.image_base64)
        if base64_size > 15_000_000:  # ~10MB after base64 encoding
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Image too large. Maximum size is approximately 10MB.",
            )

        # Get the vision service
        try:
            vision_service = get_vision_service(provider)
            logger.info(f"Using vision provider: {provider}")
        except ValueError as provider_error:
            logger.error(f"Provider error: {str(provider_error)}")
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=str(provider_error),
            )

        # Check API key configuration
        if provider == "anthropic" and not settings.ANTHROPIC_API_KEY:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Anthropic API key not configured. Please set the ANTHROPIC_API_KEY environment variable.",
            )
        elif provider == "openai" and not settings.OPENAI_API_KEY:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="OpenAI API key not configured. Please set the OPENAI_API_KEY environment variable.",
            )

        # Perform the analysis and translation
        try:
            logger.info(f"Starting vision analysis - target: {request.target_platform}, framework: {request.framework}")
            result = await vision_service.analyze_and_translate(request)
            logger.info(f"Vision analysis complete - tokens used: {result.usage.get('total_tokens', 'N/A')}")

            return ImageToPromptResponse(
                success=True,
                analysis=result.analysis,
                prompt_package=result.prompt_package,
                raw_markdown=result.raw_markdown,
                usage=result.usage,
            )

        except Exception as analysis_error:
            logger.error(f"Vision analysis error: {str(analysis_error)}", exc_info=True)
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Vision analysis failed: {str(analysis_error)}",
            )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Unexpected error in vision translate: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Unexpected error: {str(e)}",
        )


@router.post("/analyze", response_model=AnalyzeDesignResponse)
async def analyze_design(
    request: AnalyzeDesignRequest,
    provider: Literal["anthropic", "openai"] = "anthropic",
    credentials: HTTPAuthorizationCredentials = Depends(security),
    auth_service: SupabaseAuthService = Depends(get_auth_service),
):
    """
    Analyze a design image without generating prompts.

    This is a lighter-weight endpoint that only extracts design analysis
    (layout, colors, typography, components) without generating the full
    prompt package. Useful for quick previews or when you only need
    the design specifications.
    """
    try:
        # Validate user authentication
        try:
            user = await auth_service.get_user(credentials.credentials)
            logger.info(f"User authenticated for design analysis: {user.email if hasattr(user, 'email') else 'Unknown'}")
        except Exception as auth_error:
            logger.error(f"Authentication error: {str(auth_error)}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Authentication failed: {str(auth_error)}",
                headers={"WWW-Authenticate": "Bearer"},
            )

        # Convert to full request for the service
        full_request = ImageToPromptRequest(
            image_base64=request.image_base64,
            image_type=request.image_type,
            target_platform=OutputTarget.GENERIC,
            framework="react",
            styling="tailwind",
            include_interactions=False,
            detail_level="basic",  # Use basic for faster analysis
        )

        # Get the vision service
        try:
            vision_service = get_vision_service(provider)
        except ValueError as provider_error:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=str(provider_error),
            )

        # Perform analysis
        result = await vision_service.analyze_and_translate(full_request)

        return AnalyzeDesignResponse(
            success=True,
            analysis=result.analysis,
            usage=result.usage,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Unexpected error in design analysis: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Unexpected error: {str(e)}",
        )


@router.get("/platforms")
async def get_supported_platforms():
    """
    Get list of supported output platforms and their configurations.
    """
    return {
        "platforms": [
            {
                "id": "vercel_v0",
                "name": "Vercel v0",
                "description": "Optimized for v0.dev generative UI",
                "recommended_model": "claude-sonnet-4-5-20250929",
                "features": ["React", "Tailwind CSS", "shadcn/ui"],
            },
            {
                "id": "claude_code",
                "name": "Claude Code",
                "description": "Optimized for Claude Code assistant",
                "recommended_model": "claude-sonnet-4-5-20250929",
                "features": ["Full codebase context", "File operations", "Multi-file generation"],
            },
            {
                "id": "cursor",
                "name": "Cursor",
                "description": "Optimized for Cursor AI IDE",
                "recommended_model": "claude-sonnet-4-5-20250929",
                "features": ["Inline editing", "Chat-based generation", "Codebase awareness"],
            },
            {
                "id": "generic",
                "name": "Generic",
                "description": "Works with any AI coding assistant",
                "recommended_model": "claude-sonnet-4-5-20250929",
                "features": ["Universal prompts", "Framework-agnostic"],
            },
        ],
        "frameworks": ["react", "vue", "svelte", "html"],
        "styling": ["tailwind", "css", "styled-components", "scss"],
        "detail_levels": [
            {"id": "basic", "tokens": "~2000", "description": "Quick analysis"},
            {"id": "detailed", "tokens": "~4000", "description": "Standard analysis"},
            {"id": "comprehensive", "tokens": "~8000", "description": "Deep analysis with all details"},
        ],
    }
